### ISSUE TITLE
when should the action fail?

### ISSUE THREAD

**Issue by sed-i** [canonical/charm-lib-checker-action#8]
## Option 1
A change to a charm's own lib(s) sould fail the test, which means the author needs to publish the lib before merging.
In this case a PR would have implicit stages: (1) wait for integration tests to complete and the PR to be approved; (2) publish the charm's own lib(s) to chramhub and rerun this action; (3) merge.

## Option 2
>  The check should not fail for libs owned by the repo. -- @mmanciop on [prometheus-k8s-operator/189](https://github.com/canonical/prometheus-k8s-operator/pull/189#pullrequestreview-833744391)

I this case the action should probably take input args for excluding whatever needs to be excluded.
With this option we would rely on another CI action that would automatically publish libs on merge.

---

Also, if there is a long enough period of time between the last commit and the merge action, then charmlibs could be again outdated. Not sure if we want to also run this after merge as well.

**Comment by mmanciop** [comment]
> I this case the action should probably take input args for excluding whatever needs to be excluded.
With this option we would rely on another CI action that would automatically publish libs on merge.

Not really. The charm name in `metadata.yaml` matchers the charm name in the library. I am positive it can be 100% automated.

**Comment by sed-i** [comment]
> The charm name in `metadata.yaml` matchers the charm name in the library. I am positive it can be 100% automated.

Right.

Are we in agreement on having charm's own libs automatically published when charm's code merged?
I.e. that there is no use case for merging new lib code without publishing the lib.

**Comment by mmanciop** [comment]
> Are we in agreement on having charm's own libs automatically published when charm's code merged? I.e. that there is no use case for merging new lib code without publishing the lib.

Well, I am conflicted about it. Whatever object is used by the charm from its own library will be used whether the library is updated or not. That is an argument for the "change the lib, merge the lib."

On the other hand, publishing every time the library changes in a way that is visible to integrating charms may push us to release only large PRs that touch the library, increasing the chance we get them wrong (among many other side effects). @simskij ?

**Comment by simskij** [comment]
The action should never fail. It will be one of those analyzer-charms that only **alert** you that something is wrong, not enforce you to take action. I'm fairly certain that running `charmcraft fetch-lib` without arguments from the charm root will exclude owned libraries by default, but feel free to verify @sed-i.

**Comment by simskij** [comment]
Then let's add a negative matcher not the regex checking for the charm name.

**Comment by mmanciop** [comment]
> Regarding publishing the libs, I think publishing should only be done **after** merging, ie. when we release a new edge, and at that point, it could very well be automated.

Best practice is: First lib, then charm. Because charmhub has very strong ideas about not skipping patch versions in libs.

**Comment by simskij** [comment]
> > Regarding publishing the libs, I think publishing should only be done **after** merging, ie. when we release a new edge, and at that point, it could very well be automated.
> 
> Best practice is: First lib, then charm. Because charmhub has very strong ideas about not skipping patch versions in libs.

Yeah, that does not make any of the things I wrote less valid, though. The CI workflow post merge would then be:

1. lint again
2. test again
3. publish libs
4. create tag for lib
5. publish charm
6. create tag for charm

**Comment by simskij** [comment]
Action does now exclude libs owned by the charm being tested.
